File: backend-library/src/main/java/com/example/library/entity/ServiceAppointment.java
package com.example.library.entity;

import jakarta.persistence.*;
import lombok.Data;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;

/**
 * 服务预约实体。
 * 表示读者预约到馆服务、还书审核或咨询服务的记录。
 */
@Entity
@Table(name = "service_appointments")
@Data
public class ServiceAppointment {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "appointment_id")
    private Integer appointmentId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "loan_id")
    private Loan loan;

    @Enumerated(EnumType.STRING)
    @Column(name = "service_type", nullable = false)
    private ServiceType serviceType;

    @Column(name = "scheduled_time", nullable = false)
    private LocalDateTime scheduledTime;

    @Enumerated(EnumType.STRING)
    @Column(name = "method")
    private ServiceMethod method = ServiceMethod.COUNTER;

    @Enumerated(EnumType.STRING)
    @Column(name = "status")
    private AppointmentStatus status = AppointmentStatus.PENDING;

    @Column(name = "notes")
    private String notes;

    @Column(name = "return_location", length = 120)
    private String returnLocation;

    @CreationTimestamp
    @Column(name = "create_time", nullable = false, updatable = false)
    private LocalDateTime createTime;

    @UpdateTimestamp
    @Column(name = "update_time", nullable = false)
    private LocalDateTime updateTime;

    /** 服务预约类型。 */
    public enum ServiceType {
        RETURN_BOOK, PICKUP_BOOK, CONSULTATION
    }

    /** 服务办理方式。 */
    public enum ServiceMethod {
        COUNTER, SMART_LOCKER
    }

    /** 服务预约处理状态。 */
    public enum AppointmentStatus {
        PENDING, COMPLETED, CANCELLED, MISSED
    }
}

File: release/SmartLibrarySystem_Delivery_20260320_230043/front-android/src/services/serviceAppointment.ts
/**
 * @file 服务预约服务
 * @description 封装图书馆服务预约相关的 API 调用：
 *   - createAppointment：创建服务预约
 *   - getMyAppointments：分页获取我的服务预约
 *   - cancelAppointment：取消服务预约
 *
 *   服务类型：RETURN_BOOK（还书）/ PICKUP_BOOK（取书）/ CONSULTATION（咨询）
 *   服务方式：COUNTER（柜台）/ SMART_LOCKER（智能柜）
 *   预约状态：PENDING（待处理）/ COMPLETED（已完成）/ CANCELLED（已取消）/ MISSED（未到）
 */

import type {
  ApiServiceAppointmentCreateDto,
  ApiServiceAppointmentDto,
  PageResponse,
} from "../types/api";
import { request } from "./http";

/** 前端服务预约视图模型 */
export interface ServiceAppointment {
  appointmentId: number;         // 预约 ID
  userId: number;                // 用户 ID
  username: string;              // 用户名
  userFullName?: string;         // 用户全名
  loanId?: number;               // 关联的借阅 ID（还书/取书时有值）
  bookTitle?: string;            // 相关图书标题
  serviceType: "RETURN_BOOK" | "PICKUP_BOOK" | "CONSULTATION";  // 服务类型
  method: "COUNTER" | "SMART_LOCKER";                             // 服务方式
  status: "PENDING" | "COMPLETED" | "CANCELLED" | "MISSED";      // 预约状态
  scheduledTime: string;         // 预约时间
  returnLocation?: string;       // 归还地点
  notes?: string;                // 备注
  createTime: string;            // 创建时间
  updateTime?: string;           // 更新时间
}

/** 将后端服务预约 DTO 映射为前端视图模型 */
function mapAppointment(dto: ApiServiceAppointmentDto): ServiceAppointment {
  return {
    appointmentId: dto.appointmentId,
    userId: dto.userId,
    username: dto.username,
    userFullName: dto.userFullName,
    loanId: dto.loanId,
    bookTitle: dto.bookTitle,
    serviceType: dto.serviceType,
    method: dto.method,
    status: dto.status,
    scheduledTime: dto.scheduledTime,
    returnLocation: dto.returnLocation,
    notes: dto.notes,
    createTime: dto.createTime,
    updateTime: dto.updateTime,
  };
}

/** 服务预约服务对象 */
export const serviceAppointmentService = {
  /** 创建服务预约 */
  async createAppointment(
    payload: ApiServiceAppointmentCreateDto,
  ): Promise<ServiceAppointment> {
    const response = await request<ApiServiceAppointmentDto, ApiServiceAppointmentCreateDto>({
      url: "/service-appointments",
      method: "POST",
      data: payload,
      auth: true,
    });

    return mapAppointment(response);
  },

  /** 分页获取我的服务预约 */
  async getMyAppointments(page = 0, size = 20): Promise<PageResponse<ServiceAppointment>> {
    const response = await request<PageResponse<ApiServiceAppointmentDto>>({
      url: "/service-appointments/me",
      query: { page, size },
      auth: true,
    });

    return {
      ...response,
      content: (response.content ?? []).map(mapAppointment),
    };
  },

  /** 取消服务预约 */
  async cancelAppointment(appointmentId: number): Promise<void> {
    await request<void>({
      url: `/service-appointments/${appointmentId}/cancel`,
      method: "PUT",
      auth: true,
    });
  },
};
